### PR TITLE
component: channel selector for keyword detection use case

### PR DESCRIPTION
**Pull request by lbetlej**
The component allows to select from a topology file one of channels
from an input buffer to be copied to an output buffer (provides channels
count reduction functionality). In case the component is configured to
output 2 or 4 channels it works in a bypass mode.

Signed-off-by: Lech Betlej <lech.betlej@linux.intel.com>

### CODE REVIEW

File: src/audio/selector.c [hunk 1]
```diff
@@ -0,0 +1,445 @@
...
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Author: Lech Betlej <lech.betlej@linux.intel.com>
+ */
+
+/**
+ * \file audio/selector.c
+ * \brief Audio channel selection component. In case 1 output channel is
+ * \brief selected in topology the component provides the selected channel on
+ * \brief ouput. In case 2 or 4 channels are selected on output the componet
```
Comment (plbossart) [inline src/audio/selector.c]: typo: component

File: src/audio/selector.c [hunk 2]
```diff
@@ -0,0 +1,445 @@
...
+ * \param[in,out] cd Selector component private data.
+ * \param[in] in_channels Number of input channels to the module
+ * \param[in] out_channels Number of output channels to the module
+ * \param[in] ch_idx Index of a channel to be provided on output
+ */
+static void sel_set_channel_values(struct comp_data *cd, 
+						uint32_t in_channels, uint32_t out_channels, uint32_t ch_idx)
+{
+	if ((in_channels != SEL_IN_2CH) || (in_channels != SEL_IN_4CH))
+	{
```
Comment (plbossart) [inline src/audio/selector.c]: C-style? brackers are supposed to be at the end of line

File: src/audio/selector.c [hunk 3]
```diff
@@ -0,0 +1,445 @@
...
+#include <sof/work.h>
+#include <sof/clk.h>
+#include <sof/ipc.h>
+#include "selector.h"
+#include <sof/math/numbers.h>
+
+/**
+ * \brief Validates and sets channel count and channel index.
+ * \details if input data is not supported trace error is displayed and 
+ * \details input data is modified to safe values
```
Comment (plbossart) [inline src/audio/selector.c]: when the mode isn't supported, should we try to modify things or just throw an error and stop?

File: src/audio/selector.c [hunk 4]
```diff
@@ -0,0 +1,445 @@
...
+		rfree(dev);
+		return NULL;
+	}
+
+	comp_set_drvdata(dev, cd);
+
+	/* verification of initial parameters */
+	sel_set_channel_values(cd, ipc_sel->input_channels_count,
+										ipc_sel->output_channels_count, ipc_sel->selected_channel);
+
```
Comment (plbossart) [inline src/audio/selector.c]: tabs?

File: src/audio/selector.c [hunk 5]
```diff
@@ -0,0 +1,445 @@
...
+ * All done in prepare() since we need to know source and sink component params.
+ */
+static int selector_params(struct comp_dev *dev)
+{
+	trace_selector("selector_params()");
+
+	/* all configuration work is done in prepare */
+
+	return 0;
+}
```
Comment (plbossart) [inline src/audio/selector.c]: probably remove this then?

File: src/audio/selector.h [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+ */
+#define SEL_IN_2CH 2
+#define SEL_IN_4CH 4
+
+/**
+ * \brief Supported channel count on input
+ */
+#define SEL_OUT_1CH 1
+#define SEL_OUT_2CH 2
+#define SEL_OUT_4CH 4
```
Comment (plbossart) [inline src/audio/selector.h]: can't we make this more generic and support selection of M channels in N, ideally with a mask?

File: src/platform/icelake/include/platform/memory.h [hunk 7]
```diff
@@ -265,7 +265,7 @@
 /* text and data share the same HP L2 SRAM on Icelake */
 #define SOF_TEXT_START		0xBE040400
 #define SOF_TEXT_BASE		(SOF_TEXT_START)
-#define SOF_TEXT_MIN_SIZE	(0x1B000 - 0x400)
+#define SOF_TEXT_MIN_SIZE	(0x1B000 - 0x400 + 0x1000)
```
Comment (plbossart) [inline src/platform/icelake/include/platform/memory.h]: Every time we add a component, which may or may not be selected by Kconfigs, we manually tweak the memory sizes. Doesn't look scalable to me.

File: test/cmocka/src/audio/selector/selector_test.c [hunk 8]
```diff
@@ -0,0 +1,342 @@
+/*
```
Comment (plbossart) [inline test/cmocka/src/audio/selector/selector_test.c]: put the cmocka tests in separate patch (in the same PR), just to spit code and test?

File: src/audio/CMakeLists.txt [hunk 9]
```diff
@@ -16,6 +16,8 @@ if(NOT BUILD_HOST)
 		pipeline_static.c
 		component.c
 		buffer.c
+		selector.c
```
Comment (lgirdwood) [inline src/audio/CMakeLists.txt]: channel_selector.c ?

File: src/audio/selector.h [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+/** \brief Selector trace value function. */
+#define tracev_selector(__e, ...)	tracev_event(TRACE_CLASS_SELECTOR, __e, ##__VA_ARGS__)
+
+/** \brief Selector trace error function. */
+#define trace_selector_error(__e, ...)	trace_error(TRACE_CLASS_SELECTOR, __e, ##__VA_ARGS__)
+
+/**
+ * \brief Supported channel count on input
+ */
+#define SEL_IN_2CH 2
```
Comment (lgirdwood) [inline src/audio/selector.h]: can we use SOURCE and SINK terminology as this used elsewhere

File: src/audio/selector_generic.c [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+ * \brief Audio channel selector / extractor - generic processing functions
+ * \authors Lech Betlej <lech.betlej@linux.intel.com>
+ */
+
+#include "selector.h"
+
+/** 
+ * \brief channel selection for 16bit data format
+ */
+static void sel_2b_1ch(struct comp_dev *dev, struct comp_buffer *sink,
```
Comment (lgirdwood) [inline src/audio/selector_generic.c]: fwiw, we are following ALSA format naming conventions so this would be sel_s16le_1ch()

File: src/include/uapi/ipc/topology.h [hunk 12]
```diff
@@ -64,6 +64,7 @@ enum sof_comp_type {
 	SOF_COMP_EQ_FIR,
 	SOF_COMP_FILEREAD,	/**< host test based file IO */
 	SOF_COMP_FILEWRITE,	/**< host test based file IO */
+	SOF_COMP_SELECTOR,
```
Comment (lgirdwood) [inline src/include/uapi/ipc/topology.h]: The driver will just see this a MUX type (since we are switching channel source and sinks). I've added a new "flavor" to ipc_comp_confg that allows FW to sub categorise the common components like MUX, mixer, volume etc into use case specific versions of the common components to save space, PM etc. @ranj063 is taken over this work atm

File: src/include/uapi/ipc/topology.h [hunk 13]
```diff
@@ -162,6 +163,18 @@ struct sof_ipc_comp_volume {
 	uint32_t initial_ramp;	/**< ramp space in ms */
 } __attribute__((packed));
 
+/* generic selector component */
+struct sof_ipc_comp_selector {
```
Comment (lgirdwood) [inline src/include/uapi/ipc/topology.h]: This should all use the mux ipc config and source/sink/selected channels should be set at runtime via kcontrol. @ranj063 is integrating this part too.

File: src/platform/apollolake/include/platform/memory.h [hunk 14]
```diff
@@ -303,7 +303,7 @@
 
 #define SOF_TEXT_START		(HP_SRAM_VECBASE_RESET + 0x400)
 #define SOF_TEXT_BASE		(SOF_TEXT_START)
-#define SOF_TEXT_MIN_SIZE	(0x19000 - 0x400)
+#define SOF_TEXT_MIN_SIZE	(0x19000 - 0x400 + 0x1000)
```
Comment (lgirdwood) [inline src/platform/apollolake/include/platform/memory.h]: We should use Kconfig here and only add if selected. Pls see @jajanusz for info. It's working well now.

File: src/audio/CMakeLists.txt (hunk 9 above)
Comment (dbaluta) [reply src/audio/CMakeLists.txt]: @lbetlej The commit message should contain a verb. So, basically your patch adds support for a channel selector? 

Then the commit message should be:

component: Add support for channel selector

File: src/audio/selector.c (hunk 1 above)
Comment (dbaluta) [reply src/audio/selector.c]: Can you add a pseudo topology file example in the commit message. It would easier to understand how channel selector will be used.

File: src/audio/selector.c [hunk 15]
```diff
@@ -0,0 +1,445 @@
...
+ * \brief Validates and sets channel count and channel index.
+ * \details if input data is not supported trace error is displayed and 
+ * \details input data is modified to safe values
+ * \param[in,out] cd Selector component private data.
+ * \param[in] in_channels Number of input channels to the module
+ * \param[in] out_channels Number of output channels to the module
+ * \param[in] ch_idx Index of a channel to be provided on output
+ */
+static void sel_set_channel_values(struct comp_data *cd, 
+						uint32_t in_channels, uint32_t out_channels, uint32_t ch_idx)
```
Comment (dbaluta) [inline src/audio/selector.c]: Indentation of the second line of sel_set_channel_values function looks strange. Can you fix it? Look how the existing functions that split on multiple lines are indented.

File: src/audio/selector.c [hunk 16]
```diff
@@ -0,0 +1,445 @@
...
+/**
+ * \brief Sets Selector control command.
+ * \param[in,out] dev Selector base component device.
+ * \param[in,out] cdata Control command data.
+ * \return Error code.
+ */
+static int selector_ctrl_set_cmd(struct comp_dev *dev,
+			       struct sof_ipc_ctrl_data *cdata)
+{
+	//TODO add implementation aligned with with SW assumptions
```
Comment (dbaluta) [inline src/audio/selector.c]: Use C style comments /* TODO: xxx */. Anyhow, I would advise on removing empty function form the initial commit.

File: src/audio/selector.h [hunk 17]
```diff
@@ -0,0 +1,110 @@
...
+#define SELECTOR_H
+
+#include <stdint.h>
+#include <sof/audio/component.h>
+#include <sof/audio/pipeline.h>
+#include <sof/audio/format.h>
+
+#if defined(__XCC__)
+#include <xtensa/config/core-isa.h>
+#endif
```
Comment (dbaluta) [inline src/audio/selector.h]: Out of curiosity, why is this include needed?

File: src/include/uapi/ipc/topology.h [hunk 18]
```diff
@@ -162,6 +163,18 @@ struct sof_ipc_comp_volume {
...
+	struct sof_ipc_comp comp;
+	struct sof_ipc_comp_config config;
+	/* selector supports 1 input and 1 output */
+	uint32_t input_channels_count; /**< accepted values 2 or 4 */
+	uint32_t output_channels_count; /**< accepted values 1 or 2 or 4 */
+	/* note: if 2 or 4 output channels selected the component works in a bypass 
+	 * mode*/
+	uint32_t selected_channel; /**< 0..3 */
```
Comment (ranj063) [inline src/include/uapi/ipc/topology.h]: @lbetlej what does bypass mode mean? is it the same as passthrough mode? 
Then is the selected_channel valid only if the output_channels_count is 1?

File: src/audio/selector.c (hunk 3 above)
Comment (jajanusz) [reply src/audio/selector.c]: I think we should return error/fail, to be consistent - DMIC params setter for example does that if you give not supported values

File: src/platform/icelake/include/platform/memory.h (hunk 7 above)
Comment (jajanusz) [reply src/platform/icelake/include/platform/memory.h]: I'm not sure what you mean, but I think that 0x1000 should be at least define like other components that is set to 0x1000 if CONFIG_SELECTOR is here or 0x0. TBH we should not care about these TEXT sizes at all and it should just take as least as possible, calculated by tools not humans/defines.

Also when we continue having size per component you we will waste space - because we have to add size in pages - so for example component that takes 100 bytes have to take 1 page instead of just 100 bytes and when you have 2 components like that you will waste 2 pages with these defines etc...

File: src/platform/apollolake/include/platform/memory.h (hunk 14 above)
Comment (jajanusz) [reply src/platform/apollolake/include/platform/memory.h]: It works but is wasting memory, it's already explained in this comment - https://github.com/thesofproject/sof/pull/993#discussion_r255728659

File: src/audio/selector.c (hunk 2 above)
Comment (singalsu) [reply src/audio/selector.c]: The inner parentheses are not needed. Didn't the pre- and post-commit check scripts note that?

File: src/audio/selector.c [hunk 19]
```diff
@@ -0,0 +1,445 @@
...
+ * \param[in] out_channels Number of output channels to the module
+ * \param[in] ch_idx Index of a channel to be provided on output
+ */
+static void sel_set_channel_values(struct comp_data *cd, 
+						uint32_t in_channels, uint32_t out_channels, uint32_t ch_idx)
+{
+	if ((in_channels != SEL_IN_2CH) || (in_channels != SEL_IN_4CH))
+	{
+		trace_selector_error("sel_set_channel_values() error: in_channels = %u",
+		in_channels);
```
Comment (singalsu) [inline src/audio/selector.c]: Indent

File: src/audio/selector.c [hunk 20]
```diff
@@ -0,0 +1,445 @@
...
+		cd->in_channels_count = SEL_IN_2CH;
+	}
+	
+	if ((out_channels != SEL_OUT_2CH) || (in_channels != SEL_OUT_4CH) ||
```
Comment (singalsu) [inline src/audio/selector.c]: Please check commit checker script output.

File: src/audio/selector.c (hunk 2 above)
Comment (jajanusz) [reply src/audio/selector.c]: @singalsu I think checkpatch doesn't check that, anyway even when parentheses are not mandatory they improve readability imho. You are sure what order is instead of wondering what operator had what priority ;p

File: src/audio/selector.c [hunk 21]
```diff
@@ -0,0 +1,445 @@
...
+		return -EIO;	/* xrun */
+	}
+	if (sink->free < cd->sink_period_bytes) {
+		trace_selector_error("selector_copy() error: "
+						"sink component buffer"
+						" has not enough free bytes for copy");
+		comp_overrun(dev, sink, cd->sink_period_bytes, 0);
+		return -EIO;	/* xrun */
+	}
+
```
Comment (singalsu) [inline src/audio/selector.c]: You have some extra linefeeds. The pre- and post commit checker helps to find them.

File: src/audio/selector.c [hunk 22]
```diff
@@ -0,0 +1,445 @@
...
+	trace_selector("selector_new()");
+
+	if (IPC_IS_SIZE_INVALID(ipc_sel->config)) {
+		IPC_SIZE_ERROR_TRACE(TRACE_CLASS_SELECTOR, ipc_sel->config);
+		return NULL;
+	}
+
+	dev = rzalloc(RZONE_RUNTIME, SOF_MEM_CAPS_RAM,
+		COMP_SIZE(struct sof_ipc_comp_selector));
+	if (dev == NULL)
```
Comment (mrajwa) [inline src/audio/selector.c]: The checkpatch probably suggested just *(!dev)* here.

File: src/audio/selector_generic.c [hunk 23]
```diff
@@ -0,0 +1,174 @@
...
+	struct comp_buffer *source)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+	int16_t *src = (int16_t *)source->r_ptr;
+	int16_t *dest = (int16_t *)sink->w_ptr;
+	int i = 0;
+	int j = 0;
+	int nch = cd->in_channels_count; 
+	for (i = 0; i < dev->frames * nch; i += nch) {
+		dest[j] = src[i + cd->sel_channel];
```
Comment (singalsu) [inline src/audio/selector_generic.c]: You could offset the for loop with cd->sel_channel to avoid this sum operation.

File: src/audio/selector_generic.c [hunk 24]
```diff
@@ -0,0 +1,174 @@
...
+ */
+static void sel_2b_2ch(struct comp_dev *dev, struct comp_buffer *sink,
+						struct comp_buffer *source)
+{
+	int16_t *src = (int16_t *)source->r_ptr;
+	int16_t *dest = (int16_t *)sink->w_ptr;
+	int32_t i;
+
+	//TODO: change to more efficient implementation
+	for (i = 0; i < dev->frames * 2; i += 2) {
```
Comment (singalsu) [inline src/audio/selector_generic.c]: memcpy()?

File: src/audio/selector.c (hunk 1 above)
Comment (lbetlej) [reply src/audio/selector.c]: typo corrected, @dbaluta use case definition: in case keyword detector (KD) accepts mono signal and input mic array has multiple elements typically there is a need to select (in example via attribute in a topology file) the input channel for the KD, The selection is platform specific.

File: src/audio/selector.c (hunk 2 above)
Comment (lbetlej) [reply src/audio/selector.c]: breakers corrected,  I left parenthesis in place

File: src/audio/selector.c (hunk 3 above)
Comment (lbetlej) [reply src/audio/selector.c]: I've changed behavior from inform & adapt to stop-on-error

File: src/audio/selector.c (hunk 4 above)
Comment (lbetlej) [reply src/audio/selector.c]: fixed

File: src/audio/selector.h (hunk 6 above)
Comment (lbetlej) [reply src/audio/selector.h]: the component is written to enable certain use case in a finite time, from what I got from driver is that due to limitation in topology syntax there will be enum value passed from topology file to the component to select output channel, support for 2,4 channels in a pass through mode is added as a safety mechanism.  The component can be made more generic as a next step improvement.

File: src/audio/selector.h (hunk 10 above)
Comment (lbetlej) [reply src/audio/selector.h]: done

File: src/platform/icelake/include/platform/memory.h (hunk 7 above)
Comment (lbetlej) [reply src/platform/icelake/include/platform/memory.h]: I second @jajanusz regarding long term approach , for now I've added the selector component support in memory.h same as it is done for SRC for example.

File: src/include/uapi/ipc/topology.h (hunk 12 above)
Comment (lbetlej) [reply src/include/uapi/ipc/topology.h]: I plan to make this change once @ranj063 PR#1004 is accepted & merged. (part of PR#993 rebase)

File: src/include/uapi/ipc/topology.h (hunk 18 above)
Comment (lbetlej) [reply src/include/uapi/ipc/topology.h]: 2 x yes , I've updated the comment

File: src/audio/selector.c (hunk 19 above)
Comment (lbetlej) [reply src/audio/selector.c]: fixed

File: src/audio/selector.c (hunk 16 above)
Comment (lbetlej) [reply src/audio/selector.c]: TODO style comments fixed,

File: src/platform/apollolake/include/platform/memory.h (hunk 14 above)
Comment (lbetlej) [reply src/platform/apollolake/include/platform/memory.h]: kconfig fixed

File: src/platform/icelake/include/platform/memory.h (hunk 7 above)
Comment (lgirdwood) [reply src/platform/icelake/include/platform/memory.h]: @lbetlej @jajanusz yes long term is to remove, I cant quite remember why it was initially needed especially when IRAM/DRAM is separated physically on BYT anyway.

Comment (jajanusz) [reply src/platform/icelake/include/platform/memory.h]: > @lbetlej @jajanusz yes long term is to remove, I cant quite remember why it was initially needed especially when IRAM/DRAM is separated physically on BYT anyway.

Also it looks like it works only on APL because it's in base_module.c for apl and for others it's 0 and rimage can handle it

File: src/include/sof/audio/component.h [hunk 25]
```diff
@@ -512,6 +512,7 @@ static inline void sys_comp_src_init(void) {}
 void sys_comp_tone_init(void);
 void sys_comp_eq_iir_init(void);
 void sys_comp_eq_fir_init(void);
+void sys_comp_selector_init(void);
```
Comment (tlauda) [inline src/include/sof/audio/component.h]: Not called anywhere.

File: src/audio/CMakeLists.txt [hunk 26]
```diff
@@ -32,6 +32,12 @@ if(NOT BUILD_HOST)
 			src_hifi3.c
 		)
 	endif()
+	if(CONFIG_COMP_SEL)
```
Comment (tlauda) [inline src/audio/CMakeLists.txt]: You also need to add it to Kconfig file.

Comment (lbetlej) [reply src/audio/CMakeLists.txt]: Thx, fixed

File: src/include/sof/audio/component.h (hunk 25 above)
Comment (lbetlej) [reply src/include/sof/audio/component.h]: Thx, fixed

File: src/audio/selector_generic.c (hunk 24 above)
Comment (lbetlej) [reply src/audio/selector_generic.c]: thx, fixed

File: src/include/uapi/ipc/topology.h (hunk 12 above)
Comment (lgirdwood) [reply src/include/uapi/ipc/topology.h]: ok, I will merge after #1004 is resolved.

File: src/platform/icelake/include/platform/memory.h (hunk 7 above)
Comment (mmaka1) [reply src/platform/icelake/include/platform/memory.h]: That is right, almost every section that targets hpsram memory area (text/rodata/data/bss+heaps) may be packed into a single big section defined in MEMORY{} (see my req for changes in #991). It easy and quick to change and seems to be a short term must-have for apl since there are a lot of gaps hard to control manually.

File: src/audio/selector.c [hunk 27]
```diff
@@ -0,0 +1,448 @@
...
+				struct comp_buffer, source_list);
+
+	/* get source data format */
+	comp_set_period_bytes(sourceb->source, dev->frames, &cd->source_format,
+			      &cd->source_period_bytes);
+
+	/* get sink data format */
+	comp_set_period_bytes(sinkb->sink, dev->frames, &cd->sink_format,
+			      &cd->sink_period_bytes);
+
```
Comment (keyonjie) [inline src/audio/selector.c]: What are these 2 comp_set_period_bytes() calling really want to do here?
1. They are commented as "get xxx format" but actually SET period bytes and format.
2. Does it really work to set the sink component to work in 1 channel mode? The comp_set_period_bytes() invoking actually just use dev->params.channels of the sinkb->sink component directlly?

File: src/include/uapi/ipc/topology.h (hunk 18 above)
[the remark this answers is not in the document]
Comment (tlauda) [reply src/include/uapi/ipc/topology.h]: No, currently only 1 channel on output. Setting 2 or 4 will work as passthrough.

File: src/audio/selector.h (hunk 17 above)
Comment (tlauda) [reply src/audio/selector.h]: Removed.

File: src/audio/selector.c (hunk 27 above)
Comment (tlauda) [reply src/audio/selector.c]: For capture pipelines params will go upstream, so it should work correctly.
Added rewriting number of channels in selector_params to correctly propagate them further.

File: src/audio/selector.c [hunk 28]
```diff
@@ -0,0 +1,464 @@
...
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Author: Lech Betlej <lech.betlej@linux.intel.com>
+ */
+
+/**
+ * \file audio/selector.c
+ * \brief Audio channel selection component. In case 1 output channel is
+ * \brief selected in topology the component provides the selected channel on
```
Comment (mrajwa) [inline src/audio/selector.c]: Only one \brief and then continue with description?

File: src/include/uapi/ipc/topology.h [hunk 29]
```diff
@@ -162,6 +163,21 @@ struct sof_ipc_comp_volume {
...
 } __attribute__((packed));
 
+/* generic selector component */
+struct sof_ipc_comp_selector {
+	struct sof_ipc_comp comp;
+	struct sof_ipc_comp_config config;
+
+	/* selector supports 1 input and 1 output */
+	uint32_t input_channels_count;	/**< accepted values 2 or 4 */
+	uint32_t output_channels_count;	/**< accepted values 1 or 2 or 4 */
```
Comment (lgirdwood) [inline src/include/uapi/ipc/topology.h]: The limitations for channels count should be removed here (since the selector could be updated to support more). Any request for > channels than supported should return an error at topology load time.
Selecting source and sink channels should be done using masks as I dont see how the current IPC structure would allow mapping 4 source -> 2 sink channels.

**Comment by mmaka1** [comment]
@lbetlej Pls rebase

**Comment by mmaka1** [comment]
KW scan looks good, propagation of the result failed somehow.
